srt-protocol/src/connection.rs

```diff
@@ -49,3 +56,378 @@ pub struct ConnectionSettings {
+            status: ConnectionStatus::Open(settings.send_tsbpd_latency),
+            settings: settings.clone(),
+            receiver: Receiver::new(settings.clone(), connection.handshake.clone()),
+            sender: Sender::new(settings, connection.handshake),
+        }
+    }
+
+    pub fn handle_input(&mut self, now: Instant, input: Input) -> Action {
+        debug!(
+            "{:?}|{:?}|input - {:?}",
+            TimeSpan::from_interval(self.settings.socket_start_time, now),
+            self.settings.local_sockid,
+            input
+        );
+
+        match input {
+            Input::Data(Some(data)) => self.sender.handle_data(data, now),
+            Input::Data(None) => self.handle_data_stream_close(now),
+            Input::Packet(Some(packet)) => self.handle_packet(now, packet),
+            Input::Packet(None) => self.handle_socket_close(),
+            _ => {}
+        };
+
+        let action = if !self.ensure_open(now) {
+            Action::Close
+        } else if let Some(data) = self.next_data(now) {
+            Action::ReleaseData(data)
+        } else if let Some(packet) = self.next_packet() {
+            Action::SendPacket(packet)
+        } else {
+            Action::WaitForData(self.next_timer(now) - now)
+        };
+
+        debug!(
+            "{:?}|{:?}|action - {:?}",
+            TimeSpan::from_interval(self.settings.socket_start_time, now),
+            self.settings.local_sockid,
+            action
+        );
+
+        action
+    }
+
+    pub fn is_open(&self) -> bool {
+        !(self.status == ConnectionStatus::Closed
+            || (!self.sender.is_open() && self.receiver.is_flushed())
+            || (!self.receiver.is_open() && self.sender.is_flushed()))
+    }
+
+    pub fn next_packet(&mut self) -> Option<(Packet, SocketAddr)> {
+        self.sender
+            .next_packet()
+            .or_else(|| self.receiver.next_packet())
+    }
+
+    pub fn next_data(&mut self, now: Instant) -> Option<(Instant, Bytes)> {
+        self.receiver.next_data(now)
+    }
+
+    pub fn next_timer(&self, now: Instant) -> Instant {
+        min(self.sender.next_timer(now), self.receiver.next_timer(now))
+    }
+
+    pub fn ensure_open(&mut self, now: Instant) -> bool {
+        if !self.is_open() {
+            false
+        } else {
+            self.check_timers(now);
+            true
+        }
+    }
+
+    pub fn check_timers(&mut self, now: Instant) -> Instant {
+        let _status = &mut self.status;
+        let sender = &mut self.sender;
+        let receiver = &mut self.receiver;
+
+        sender.check_timers(now);
+        receiver.check_timers(now);
+
+        // if status.check_shutdown(now, || sender.is_flushed()) {
+        //     debug!("{:?} sending shutdown", self.settings.local_sockid);
+        //     sender.send_shutdown(now);
+        // }
+
+        self.next_timer(now)
+    }
+
+    pub fn handle_data_input(&mut self, now: Instant, data: Option<(Instant, Bytes)>) {
```

Comment:
Why is this not just a wrapper over handle_input or vice versa?

Reply:
I did this in order to differentiate between the two public API designs I was experimenting with. Action/Input vs multiple command and query approach (i.e. is_open, check_timers, handle_packet, handle_data, next_packet, next_data)